File: ios/AutoskeletonTracing.swift
import Foundation
import os

// Task 3.1/3.2 (tasks.md Phase 3) / spec.md REQ-OBS-PROFILE-1: `os_signpost`/
// `OSSignposter` intervals around traversal and draw. `OSSignposter` itself has no
// public way to assert "an interval was opened and closed with this name" from
// XCTest — Instruments capture is not scriptable in a unit test target — so this
// file defines a thin, injectable seam: production code always goes through
// `AutoskeletonTracing`, whose default implementation wraps a real `OSSignposter`;
// tests inject a recording double and assert on the recorded begin/end pairs
// instead. This is the standard pattern for testing signpost usage and is the
// "dedicated XCTest" plan.md §7.1 and §7.2 both call for.
protocol AutoskeletonTracing {
    /// Opens a named interval, returning an opaque token that must be passed back
    /// to `end(_:token:)` to close it.
    func begin(_ name: StaticString) -> Any
    func end(_ name: StaticString, token: Any)
}

/// Production implementation: a real `OSSignposter` on a dedicated `OSLog`
/// subsystem/category so traversal/draw intervals are distinguishable from system
/// noise in Instruments' Points of Interest track.
final class AutoskeletonSignpostTracing: AutoskeletonTracing {
    private let signposter: OSSignposter

    init(subsystem: String = "com.autoskeleton", category: String = "AutoskeletonSensor") {
        signposter = OSSignposter(logger: Logger(subsystem: subsystem, category: category))
    }

    func begin(_ name: StaticString) -> Any {
        let id = signposter.makeSignpostID()
        return (id, signposter.beginInterval(name, id: id))
    }

    func end(_ name: StaticString, token: Any) {
        guard let (id, state) = token as? (OSSignpostID, OSSignpostIntervalState) else { return }
        signposter.endInterval(name, state)
        _ = id
    }
}

/// Test double: records every begin/end call, in order, so a test can assert both
/// that an interval was opened AND that it was closed (not left dangling), and that
/// nesting/ordering matches REQ-OBS-PROFILE-2's same-thread begin/end discipline.
final class AutoskeletonRecordingTracing: AutoskeletonTracing {
    enum Event: Equatable {
        case begin(String)
        case end(String)
    }

    private(set) var events: [Event] = []
    private var nextToken = 0

    func begin(_ name: StaticString) -> Any {
        events.append(.begin("\(name)"))
        nextToken += 1
        return nextToken
    }

    func end(_ name: StaticString, token: Any) {
        events.append(.end("\(name)"))
    }
}

File: android/src/main/java/com/autoskeleton/AutoskeletonTracing.kt
package com.autoskeleton

import android.os.Trace

// Task 4.1 (tasks.md Phase 4) / spec.md REQ-OBS-PROFILE-1/2: `Trace.beginSection`/
// `endSection` intervals around traversal and draw. Same injectable-seam pattern as
// iOS's `AutoskeletonTracing.swift`: `Trace.beginSection`/`endSection` are
// fire-and-forget system calls with nothing to assert on directly from a test (and
// Robolectric's shadow does not expose recorded section names for inspection), so
// production code depends on this interface and tests inject a recording double
// instead.
//
// REQ-OBS-PROFILE-2 (name <= 127 chars, same-thread begin/end nesting) is enforced
// by `AutoskeletonSystemTracing.begin` itself (a hard `require`) rather than left as
// a documentation-only constraint — asserted directly by a dedicated test.
interface AutoskeletonTracing {
    fun begin(name: String)
    fun end(name: String)
}

const val AUTOSKELETON_MAX_TRACE_SECTION_NAME_LENGTH = 127

/** Production implementation: real `Trace.beginSection`/`endSection` on the calling
 *  thread — same-thread nesting (REQ-OBS-PROFILE-2) holds by construction since
 *  Android's `Trace` section stack is itself thread-local. */
class AutoskeletonSystemTracing : AutoskeletonTracing {
    override fun begin(name: String) {
        require(name.length <= AUTOSKELETON_MAX_TRACE_SECTION_NAME_LENGTH) {
            "Trace section name exceeds the $AUTOSKELETON_MAX_TRACE_SECTION_NAME_LENGTH-character " +
                "limit (REQ-OBS-PROFILE-2): $name"
        }
        Trace.beginSection(name)
    }

    override fun end(name: String) {
        Trace.endSection()
    }
}

/** Test double: records every begin/end call, in order, so a test can assert both
 *  that an interval was opened AND that it was closed (not left dangling), and that
 *  nesting/ordering matches REQ-OBS-PROFILE-2's same-thread begin/end discipline. */
class AutoskeletonRecordingTracing : AutoskeletonTracing {
    sealed class Event {
        data class Begin(val name: String) : Event()
        data class End(val name: String) : Event()
    }

    val events = mutableListOf<Event>()

    override fun begin(name: String) {
        events.add(Event.Begin(name))
    }

    override fun end(name: String) {
        events.add(Event.End(name))
    }
}
